src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,168 @@
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DocumentResolver _documentResolver;
+
+        public override string Action => Constants.ExtractToCodeBehindAction;
+
+        public ExtractToCodeBehindCodeActionResolver(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver)
+        {
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+        }
+
+        public override async Task<WorkspaceEdit> ResolveAsync(JObject data, CancellationToken cancellationToken)
+        {
+            var actionParams = data.ToObject<ExtractToCodeBehindParams>();
+            var path = Path.GetFullPath(actionParams.Uri.LocalPath);
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            if (document is null)
+            {
+                return null;
+            }
+
+            var text = await document.GetTextAsync();
+            if (text is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var codeBehindPath = "";
+            var n = 0;
+            do
+            {
+                var identifier = n > 0 ? n.ToString() : "";  // Make it look nice
+                codeBehindPath = Path.Combine(
+                    Path.GetDirectoryName(path),
+                    $"{Path.GetFileNameWithoutExtension(path)}{identifier}{Path.GetExtension(path)}.cs");
+                n++;
+            } while (File.Exists(codeBehindPath));
+
+            var codeBehindUri = new UriBuilder()
+            {
+                Scheme = Uri.UriSchemeFile,
+                Path = codeBehindPath,
+                Host = string.Empty,
+            }.Uri;
+
+            var className = Path.GetFileNameWithoutExtension(path);
+            var contents = text.GetSubTextString(new CodeAnalysis.Text.TextSpan(actionParams.ExtractStart, actionParams.ExtractEnd - actionParams.ExtractStart));
+            var compilationUnit = GenerateCodeBehindClass(className, contents, codeDocument);
```

Comment:
I'd just make the `GenerateCodeBehindClass` return a string since you toString the compilation unit anyways. That way the idea of a compilation unit doesn't bleed out of the class generation.

```suggestion
            var generatedClassContent = GenerateCodeBehindClass(className, contents, codeDocument);
```